citylearn/reward_function.py

```diff
@@ -307,4 +307,134 @@ def calculate(self, observations: List[Mapping[str, Union[int, float]]]) -> List
         reward = reward*np.reshape(self.coefficients, (len(self.coefficients), 1))
         reward = reward.sum(axis=0).tolist()
 
-        return reward
\ No newline at end of file
+        return reward
+
+
+class V2GPenaltyReward(MARL):
+    """Rewards with considerations for electric vehicle charging behaviours in a V2G setting.
+    Note that this function rewards/penalizes only the electric vehicle part. For a comprehensive reward strategy
+    please use one of the super classes or rewrite your own
+
+    Parameters
+    ----------
+    env_metadata: Mapping[str, Any]:
+        General static information about the environment.
+    """
+
+    def __init__(self, env_metadata: Mapping[str, Any],
+                 peak_percentage_threshold=0.10,
+                 ramping_percentage_threshold=0.10,
+                 peak_penalty_weight=20,
+                 ramping_penalty_weight=15,
+                 energy_transfer_bonus=10,
+                 window_size=6,
+                 penalty_no_car_charging=-5,
+                 penalty_battery_limits=-2,
+                 penalty_soc_under_5_10=-5,
+                 reward_close_soc=10,
+                 reward_self_ev_consumption=5,
+                 community_weight=0.2,
+                 reward_extra_self_production=5):
+        super().__init__(env_metadata)
+
+        # Setting the parameters
+        self.PEAK_PERCENTAGE_THRESHOLD = peak_percentage_threshold
+        self.RAMPING_PERCENTAGE_THRESHOLD = ramping_percentage_threshold
+        self.PEAK_PENALTY_WEIGHT = peak_penalty_weight
+        self.RAMPING_PENALTY_WEIGHT = ramping_penalty_weight
+        self.ENERGY_TRANSFER_BONUS = energy_transfer_bonus
+        self.WINDOW_SIZE = window_size
+        self.PENALTY_NO_CAR_CHARGING = penalty_no_car_charging
+        self.PENALTY_BATTERY_LIMITS = penalty_battery_limits
+        self.PENALTY_SOC_UNDER_5_10 = penalty_soc_under_5_10
+        self.REWARD_CLOSE_SOC = reward_close_soc
+        self.COMMUNITY_WEIGHT = community_weight
+        self.REWARD_EXTRA_SELF_PRODUCTION = reward_extra_self_production
+        self.REWARD_SELF_EV_CONSUMPTION = reward_self_ev_consumption
+
+    def calculate(self, observations: List[Mapping[str, Union[int, float]]]) -> List[float]:
+
+        ##net_electricity_consumption = [o['net_electricity_consumption'] for o in observations]
+        current_reward = super.calculate(observations)
+        print (observations)
+        reward_list = []
+
+        for b in self.env.buildings:

```

Comment:
This reward function will not work because it assumes it has access to the environment, `env`, when it only has access to the `env_metadata`  that is a `dict` of static environment data. The reason why we only provide the `env_metadata` is because we didn't want participants in the CityLearn Challenge to have full access to the environment when calculating rewards and risk mutating something they shouldn't. The only way you can infer the current state of the environment is through the `observations` variable parsed to the `calculate` method.

Reply:
Is there a different reward function I should replace it with? Or did you want me to follow up with Tiago?